opt: remove fallback to heuristic planner

RaduBerinde opened the pull request:
The optimizer is able to handle all statements now; we can remove the
logic to fall back to the heuristic planner.

A couple of things needed fixing:
 - we didn't include `Import` and `ShowBackup` as opaque statements (they
   weren't implementing `CCLOnlyStatement`, which is also fixed).
 - the HP was incorrectly hitting the CTE telemetry counter even when
   there was no `WITH`. In the existing test, the HP was running as
   fallback because of some unimplemented functions.

Release note: None

RaduBerinde committed:
opt: remove fallback to heuristic planner

The optimizer is able to handle all statements now; we can remove the
logic to fall back to the heuristic planner.

A couple of things needed fixing:
 - we didn't include Import and ShowBackup as opaque statements (they
   weren't implementing `CCLOnlyStatement`, which is also fixed).
 - the HP was incorrectly hitting the CTE telemetry counter even when
   there was no `WITH`. In the existing test, the HP was running as
   fallback because of some unimplemented functions.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/39362)
<!-- Reviewable:end -->

RaduBerinde commented:
This is a major milestone: as of this PR all statements officially go through the optimizer. We are very close to removing the heuristic planner altogether (most of the internal paths that require it have been dealt with).

jordanlewis commented:
Wow, amazing! Congratulations, opt team.

petermattis commented:
:tada: :tada: :tada:

tbg commented:
Congrats!

RaduBerinde committed:
opt: clean up isCorrelated flag

In the planning code we are plumbing through an `isCorrelated` flag
used to enhance the error message when we fall back to the heuristic
planner and the query was correlated. Since we no longer fall back,
this flag is not needed anymore.

Release note: None

RaduBerinde commented:
Added a commit to clean up the `isCorrelated` stuff.

rytaft reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 8 of 8 files at r1, 8 of 8 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39362#-:-LlgDmK59mu3cX_KgZov:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball and @justinj)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
